File: army/JavaArmy/src/com/maksklaban/armymax/spell/FireBall.java
package com.maksklaban.armymax.spell;

import com.maksklaban.armymax.units.*;
import com.maksklaban.armymax.exceptions.*;
import com.maksklaban.armymax.spellcasters.*;

public class FireBall extends Spell {
    public FireBall(String spellName, int actionPoints, int manaCost) {
        super(spellName, actionPoints, manaCost);
    }

    public FireBall() {
        this("FireBall", 40, 30);
    }

    public void cast(Spellcaster caster, Unit enemy, Spell spell) throws ArmyException {
        caster.spendManaPoints(this.getManaCost());

        if ( caster.getTitle() == "Priest" ) {
            this.priestCast(caster, enemy, spell);
        } else if ( caster.getTitle() == "Necromanser" ) {
            this.necromanserCast(caster, enemy, spell);
        } else {
            if ( caster.getIsBattle() ) {
                enemy.takeMagDamage(this.getActionPoints());
            } else {
                enemy.takeMagDamage(this.getActionPoints() / 2);
            }
        }
    }

    protected void priestCast(Spellcaster caster, Unit enemy, Spell spell) throws ArmyException {
        if ( enemy.getIsUndead() ) {
            enemy.takeMagDamage(this.getActionPoints());
        } else {
            enemy.takeMagDamage(this.getActionPoints() / 2);
        }
    }

    protected void necromanserCast(Spellcaster caster, Unit enemy, Spell spell) throws ArmyException {
        enemy.takeMagDamage(this.getActionPoints());
        enemy.attach(caster);
    }
}

File: army/Spell/FireBall.cpp
#include "FireBall.h"

FireBall::FireBall(std::string spellName, int manaCost, int actionPoints) : Spell(spellName, actionPoints, manaCost) {}

FireBall::~FireBall() {}

void FireBall::cast(Spellcaster* caster, Unit* enemy, Spell* spell) {
    caster->spendManaPoints(this->getManaCost());

    if ( caster->getTitle() == "Priest" ) {
        this->priestCast(caster, enemy, spell);
    } else if ( caster->getTitle() == "Necromanser" ) {
        this->necromanserCast(caster, enemy, spell);
    } else {
        if ( caster->getIsBattle() ) {
            enemy->takeMagDamage(spell->getActionPoints());
        } else {
            enemy->takeMagDamage(spell->getActionPoints() / 2);
            }
        }
}

void FireBall::priestCast(Spellcaster* caster, Unit* enemy, Spell* spell) {
    if ( enemy->getIsUndead() ) {
        enemy->takeMagDamage(spell->getActionPoints());
    } else {
        enemy->takeMagDamage(spell->getActionPoints() / 2);
    }
}

void FireBall::necromanserCast(Spellcaster* caster, Unit* enemy, Spell* spell) {
    enemy->takeMagDamage(spell->getActionPoints());
    caster->addSubject(enemy);
}
